Data/Vector/Unboxed/Base.hs

```diff
@@ -412,11 +421,129 @@ instance (Unbox a) => G.Vector Vector (Complex a) where
+; {-# INLINE basicSet             #-}                                         \
+; {-# INLINE basicUnsafeCopy      #-}                                         \
+; {-# INLINE basicUnsafeGrow      #-}                                         \
+; basicLength (con v)                = M.basicLength v                        \
+; basicUnsafeSlice i n (con v)       = con $ M.basicUnsafeSlice i n v         \
+; basicOverlaps (con v1) (con v2)    = M.basicOverlaps v1 v2                  \
+; basicUnsafeNew n                   = con `liftM` M.basicUnsafeNew n         \
+; basicInitialize (con v)            = M.basicInitialize v                    \
+; basicUnsafeReplicate n (tyC x)     = con `liftM` M.basicUnsafeReplicate n x \
+; basicUnsafeRead (con v) i          = tyC `liftM` M.basicUnsafeRead v i      \
+; basicUnsafeWrite (con v) i (tyC x) = M.basicUnsafeWrite v i x               \
+; basicClear (con v)                 = M.basicClear v                         \
+; basicSet (con v) (tyC x)           = M.basicSet v x                         \
+; basicUnsafeCopy (con v1) (con v2)  = M.basicUnsafeCopy v1 v2                \
+; basicUnsafeMove (con v1) (con v2)  = M.basicUnsafeMove v1 v2                \
+; basicUnsafeGrow (con v) n          = con `liftM` M.basicUnsafeGrow v n      \
+}
+#define newtypeVector(ty,tyC,con,mcon) \
+instance (Unbox a) => G.Vector Vector (ty a) where { \
+; {-# INLINE basicUnsafeFreeze  #-}                                       \
+; {-# INLINE basicUnsafeThaw    #-}                                       \
+; {-# INLINE basicLength        #-}                                       \
+; {-# INLINE basicUnsafeSlice   #-}                                       \
+; {-# INLINE basicUnsafeIndexM  #-}                                       \
+; {-# INLINE elemseq            #-}                                       \
+; basicUnsafeFreeze (mcon v)        = con `liftM` G.basicUnsafeFreeze v   \
+; basicUnsafeThaw (con v)           = mcon `liftM` G.basicUnsafeThaw v    \
+; basicLength (con v)               = G.basicLength v                     \
+; basicUnsafeSlice i n (con v)      = con $ G.basicUnsafeSlice i n v      \
+; basicUnsafeIndexM (con v) i       = tyC `liftM` G.basicUnsafeIndexM v i \
+; basicUnsafeCopy (mcon mv) (con v) = G.basicUnsafeCopy mv v              \
+; elemseq _ (tyC a)                 = G.elemseq (undefined :: Vector a) a \
+}
+
+
 #if MIN_VERSION_base(4,8,0)
 newtype instance MVector s (Identity a) = MV_Identity (MVector s a)
 newtype instance Vector (Identity a) = V_Identity (Vector a)
+instance (Unbox a) => Unbox (Identity a)
+newtypeMVector(Identity, Identity, MV_Identity)
+newtypeVector(Identity, Identity, V_Identity, MV_Identity)
+#endif
 
-instance (Unbox a) => M.MVector MVector (Identity a) where
+newtype instance MVector s (Down a) = MV_Down (MVector s a)
+newtype instance Vector (Down a) = V_Down (Vector a)
+instance (Unbox a) => Unbox (Down a)
```

Comment:
As a general thought, is it possible to have the `newtypeMVector`/`newtypeVector` macros generate these `MVector`, `Vector`, and `Unbox` instances? That might save a good deal of boilerplate.